app/scripts/components/common/chart/utils.ts

```diff
@@ -0,0 +1,64 @@
+import { timeFormat, timeParse } from 'd3-time-format';
+import * as d3ScaleChromatic from 'd3-scale-chromatic';
+
+export const dateFormatter = (date: Date, dateFormat: string) => {
+  const format = timeFormat(dateFormat);
+  return format(date);
+};
+
+export const convertToTime = ({ timeString, dateFormat}: {timeString: string, dateFormat: string}) => {
+  if (!timeString) return null;
+
+  const parseDate = timeParse(dateFormat);
+  return parseDate(timeString).getTime();
+};
+
+export function getFData({ data, idKey, xKey, yKey, dateFormat }: { data: any[], idKey: string, xKey: string, yKey: string, dateFormat: string }) {
+  const uniqueKeys = [...Array.from(new Set(data.map((d) => d[idKey])))].sort();
+
+  const fData: object[] = [];
+  data.reduce((acc, curr) => {
+    // Use acc object so we don't have to iterate (ex.Array.find) the array to find an element
+    if (!acc[curr[xKey]]) {
+      acc[curr[xKey]] = {
+        [curr[idKey]]: parseFloat(curr[yKey])
+      };
+    } else {
+      acc[curr[xKey]] = {
+        ...acc[curr[xKey]],
+        [curr[idKey]]: parseFloat(curr[yKey])
+      };
+      // Once the object got all the attributes with matching xKey, push it to final data
+      if (Object.keys(acc[curr[xKey]]).length === uniqueKeys.length) {
+        fData.push({
+          ...acc[curr[xKey]],
+          [xKey]: convertToTime({
+            timeString: curr[xKey],
+            dateFormat
+          })
+        });
+      }
+    }
+
+    return acc;
+  }, {});
```

Comment:
It is a bit complicated to understand the flow of this function, specially since a `.reduce` function is expected to return something. Can the be rewritten in a simpler way?

Reply:
I added comments, I hope it helps. The function have gotten less readable because it was trying to avoid 
 additional iterations. 

If you have a better idea how to format `[{year: 2005, no2: 10}, {year: 2005, so2: 5}] ` to `[{year:2005, no2:10, so2:5]` (key names - year(xKey), no2, so2(uniqueKeys- driven from idKey) - are dynamic, coming from data), I would love to hear!

+ The function would look like below with `find` method. 
```
  const fData = data.reduce((acc, curr) => {
    // check current xKey value is already in acc
    const currentTime =  convertToTime({
      timeString: curr[xKey],
      dateFormat
    });
    const currentXKeyObjectIdx = acc.findIndex(e => e[xKey] == currentTime);
   // if not, push with xkey along with other properties
    if (currentXKeyObjectIdx < 0) {
      acc.push({
        [xKey]: convertToTime({
          timeString: curr[xKey],
          dateFormat
        }),
      [curr[idKey]]: parseFloat(curr[yKey])
    });
  } else {
   // if there is already an object with current xkey value, replace it to an object with an additional property
    const currentXKeyObject = acc[currentXKeyObjectIdx];
    acc[currentXKeyObjectIdx] = {
      ...currentXKeyObject,
      [curr[idKey]]: parseFloat(curr[yKey])
    };
  }
  return acc;
  },[]);
```

Reply:
What I'm seeing from the data and what you said, it is a matter of grouping the data by year (the xKey), and merging the other properties.

I'd implement it as follows:
```js
// Group data by the xKey.
  const resultByXKey = data.reduce((acc, entry) => {
    // ## From:
    // {
    //   "xkey": xKey value,
    //   "yKey": yKey value,
    //   "idKey": idKey value
    // }
    // ## to
    // {
    //   "xkey": xKey value,
    //   "idKey value": yKey value
    // }
    const formattedObj = {
      // [xKey]: entry[xKey], // year not formatted
      [xKey]: convertToTime({
        timeString: entry[xKey],
        dateFormat
      }),
      [entry[idKey]]: parseFloat(entry[yKey])
    };

    const groupValue = formattedObj[xKey];

    if (acc[groupValue]) {
      return {
        ...acc,
        [groupValue]: {
          ...acc[groupValue],
          ...formattedObj
        }
      };
    } else {
      return {
        ...acc,
        [groupValue]: formattedObj
      };
    }
  }, {});
  
  const results = Object.values(resultByXKey);
```

This works well for the 1st chart, but returns different values for the 2nd.
This code returns results for the year 2022 (which are present in the source), but your code doesn't. Not sure what is expected.

Reply:
Ah it is great idea to use `Object.values`. 
I think the reason the data didn't have 2022 entries was the logic I put was eliminating the data point that doesn't have all the idKeys. (To put more simply, not all the counties have data for 2022)
I reflected your comment into the code!